pkg/controller/elasticsearch/driver/nodes.go

```diff
@@ -110,8 +111,46 @@ func (d *defaultDriver) reconcileNodeSpecs(
 		return results
 	}
 
+	// When not reconciled, set the phase to pending only if it's operational to avoid to
+	// override another "not operational" phase like MigratingData.
+	if Reconciled(expectedResources.StatefulSets(), actualStatefulSets, resourcesState) {
+		reconcileState.UpdateElasticsearchOperational(resourcesState, observedState)
+	} else if reconcileState.IsElasticsearchOperational(observedState) {
+		reconcileState.UpdateElasticsearchPending(resourcesState.CurrentPods)
+	}
+
 	// TODO:
 	//  - change budget
 	//  - grow and shrink
 	return results
 }
+
+// Reconciled reports whether the actual StatefulSets are reconciled to match the expected StatefulSets
+// by checking that the expected template hash label is reconciled for all StatefulSets and all their
+// associated pods.
+func Reconciled(expectedStatefulSets, actualStatefulSets sset.StatefulSetList, resourcesState reconcile.ResourcesState) bool {
+	for _, expectedSset := range expectedStatefulSets {
+		actualSset, ok := actualStatefulSets.GetByName(expectedSset.Name)
+		if !ok {
+			return false
+		}
+		// actual sset should have the expected sset template hash label
+		if !sset.EqualTemplateHashLabels(expectedSset, actualSset) {
+			log.V(1).Info("Statefulset not reconciled",
+				"statefulset_name", expectedSset.Name, "reason", "template hash not equal")
+			return false
+		}
+		// pods should have the expected sset template hash label
+		ssetConfigTplHash := expectedSset.Spec.Template.Labels[label.ConfigTemplateHashLabelName]
+		for _, pod := range resourcesState.AllPods {
+			if pod.Labels[label.ConfigTemplateHashLabelName] != ssetConfigTplHash {
```

Comment:
* Seems a bit strange that we check Pods are upgraded through their sset template hash here, whereas we check the [revision in another place of the code to know if a Pods should be upgraded](https://github.com/elastic/cloud-on-k8s/blob/24f1822c7302a4a8cd174fba2d07c055cb5e69f4/pkg/controller/elasticsearch/driver/upgrade.go#L188). Should we maybe just reuse [podsToUpgrade(actualStatefulSets)](https://github.com/elastic/cloud-on-k8s/blob/24f1822c7302a4a8cd174fba2d07c055cb5e69f4/pkg/controller/elasticsearch/driver/upgrade.go#L155) here? It should return an empty list of no pod needs an upgrade.

* We probably also need to check that no pod is missing (eg. in the process of being added/removed), which can be done by [checking expectations](https://github.com/elastic/cloud-on-k8s/blob/24f1822c7302a4a8cd174fba2d07c055cb5e69f4/pkg/controller/elasticsearch/driver/expectations.go#L12)


So the function would look like:

1. Do actualStatefulSets == expectedStatefulSets? If not, Pending.
1. Are expectations met? If not, Pending.
1. Are there pods to upgrade? If yes, Pending.
1. Are all pods ready? If not, Pending.
1. -> Operational.

Reply:
> Should we maybe just reuse podsToUpgrade(actualStatefulSets) here?

Yes.

> We probably also need to check that no pod is missing (eg. in the process of being added/removed), which can be done by checking expectations

Expectations are verified just above so the phase will not updated to Operational if . I thought it was enough. 
- reconciliation 1: expectations ok, `Upscale|Downscale|Upgrade` is applied, `Reconciled` returns false, => Pending
- reconciliation 2: !expectations, => requeue
- reconciliation x: expectations ok, `Reconciled` returns true, => Operational

> 1. Do actualStatefulSets == expectedStatefulSets? If not, Pending.

We can't make a strict comparison (thinking about `reflect.DeepEqual(a, e)`) because there are [too many differences](https://gist.github.com/thbkrkr/09450f03eb5385144761a26d5a5f9a4e) between the two objects. This is why I compared the template hash label like we do in [`ReconcileStatefulSet`](https://github.com/elastic/cloud-on-k8s/blob/571a78aaf8e2dbc49d839e12c48c15dde7ccb3f2/pkg/controller/elasticsearch/sset/reconcile.go#L29) and introduced a small function `EqualTemplateHashLabels` to reuse the code between the two places.